proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,515 @@
+register types are valid for that variable.
+
+## Proposed solution
+The following table lists all the proposed diagnostic
+names and diagnostic messages that are relevant in this spec:
+
+| Diagnostic Name | Diagnostic message | Description |
+|-|-|-|
+| err_hlsl_mismatching_register_type_and_variable_type | "error: unsupported resource register binding '%select{t\|u\|b\|s}0' on variable of type '%1'"| Emitted if a variable type that isn't accompanied with a resource class attribute is bound with a standard register type (`t`, `u`, `b`, or `s`). |
+| err_hlsl_unsupported_register_type_and_variable_type | "error: register binding type '%0' not supported for variable of type '%1'" | Emitted if a variable type is bound using an unsupported register type, like binding a float with the 'x' register type. |
+| err_hlsl_mismatching_register_type_and_resource_type | "error: %select{srv\|uav\|cbv\|sampler}2 type '%0' requires register type '%select{t\|u\|b\|s}2', but register type '%1' was used." | Emitted if a known resource type is bound using a standard but mismatching register type, e.g., RWBuffer<int> getting bound with 's'.|
+| err_hlsl_unsupported_register_type_and_user_defined_type | "error: invalid register type '%0' used; expected 't', 'u', 'b', or 's'"| Emitted if an unsupported register type is used to bind a UDT.|
+| err_hlsl_conflicting_register_annotations | "error: conflicting register annotations: multiple register annotations detected for register type '%0'" | Emitted if two register annotations with the same register type are applied to the same declaration. |
+| warn_hlsl_register_type_c_not_in_global_scope | "warning: register binding 'c' ignored inside cbuffer/tbuffer declarations; use pack_offset instead" | Emitted if a basic type is bound with `c` within a `cbuffer` or `tbuffer` scope |
+| warn_hlsl_deprecated_register_type_b | "warning: deprecated legacy bool constant register binding 'b' used. 'b' is only used for constant buffer resource binding." | Emitted if the register prefix `b` is used on a variable type without the resource class attribute, like a float type. |
+| warn_hlsl_deprecated_register_type_i | "warning: deprecated legacy int constant register binding 'i' used." | Emitted if the register prefix `i` is used on a variable type without the resource class attribute, like a float type. |
+| warn_hlsl_user_defined_type_missing_resource_type_member | "warning: variable of type '%0' bound to register type '%1' does not contain a matching '%select{srv\|uav\|cbv\|sampler}2' resource" | Emitted if a UDT is lacking any eligible member to bind using a specific register type, like if a UDT `Foo` was bound with `s` but had no sampler members.|
+| warn_hlsl_user_defined_type_missing_basic_type | "warning: register 'c' used on type with no contents to allocate in a constant buffer" | Emitted if a UDT has no numerical members, and is being bound using the `c` register type.|

```

Comment:
My issue with using `warn_hlsl_binding_type_mismatch` to replace `warn_hlsl_user_defined_type_missing_resource_type_member` and `warn_hlsl_user_defined_type_missing_basic_type`:
'c' is separate, because to my understanding, it isn't quite a register type, and it isn't binding a resource. So, the language used for 'c', "constant buffer data", should apply to register type 'b'. 'c' is for basic types that are to be placed in the $Globals constant buffer, which is a separate concept when compared to allocating a separate constant buffer, which is specified by 'b'. So I'm not quite sure 'c' is a "binding type".
I could change the message to include a missing type for each of the 5 register types (t u b s c), but I think the separation of concepts is significant enough to warrant a separate diagnostic in the "c" case.

Issue with `err_hlsl_binding_type_mismatch` replacing `err_hlsl_mismatching_register_type_and_resource_type` and `err_hlsl_unsupported_register_type_and_variable_type`:
There should be separate errors to indicate whether the register type is valid or invalid, and whether the register type is valid for a specific resource or not. Your suggested replacement diagnostic doesn't mention the register binding type that was used, which I think is helpful information. I could just rename the diagnostic to `err_hlsl_binding_type_mismatch` from `err_hlsl_mismatching_register_type_and_resource_type`, but I think the existing message is better, what do you think?

My issue with `err_hlsl_binding_type_invalid` replacing `err_hlsl_unsupported_register_type_and_variable_type`:
It is important the diagnostic lists "for variable type x" instead of just declaring the register type invalid. This is because the diagnostic gets emitted not only when the register type is invalid (like "y"), but also when a completely invalid variable type is attempted to be bound (like RayQuery). We emit the diagnostic in this case, so it could be that "t" is used with RayQuery, which is a valid register type. It's just that the variable type is invalid. We could take your suggestion, but that would require adding a new diagnostic for cases where the variable type is all-around invalid. Thoughts?

I can accept the rename to `err_hlsl_duplicate_register_annotation`
I can accept the rename to `warn_hlsl_register_type_c_packoffset`
I can accept both rewording suggestions.

Reply:
> My issue with using `warn_hlsl_binding_type_mismatch` to replace `warn_hlsl_user_defined_type_missing_resource_type_member` and `warn_hlsl_user_defined_type_missing_basic_type`: 'c' is separate, because to my understanding, it isn't quite a register type, and it isn't binding a resource. So, the language used for 'c', "constant buffer data", should apply to register type 'b'. 'c' is for basic types that are to be placed in the $Globals constant buffer, which is a separate concept when compared to allocating a separate constant buffer, which is specified by 'b'. So I'm not quite sure 'c' is a "binding type". I could change the message to include a missing type for each of the 5 register types (t u b s c), but I think the separation of concepts is significant enough to warrant a separate diagnostic in the "c" case.

The way I see it, the 'b' binding type is for a constant buffer. The 'c' binding type is for "constant buffer data (in the $Globals constant buffer)". Given that 'b' bindings are *never* valid on a VarDecl this separation seemed clean to me, and I think misuses of the 'b' type are the actual outlier here (which is covered by the `warn_hlsl_deprecated_register_type_b` diagnostic).

> Issue with `err_hlsl_binding_type_mismatch` replacing `err_hlsl_mismatching_register_type_and_resource_type` and `err_hlsl_unsupported_register_type_and_variable_type`: There should be separate errors to indicate whether the register type is valid or invalid, and whether the register type is valid for a specific resource or not. Your suggested replacement diagnostic doesn't mention the register binding type that was used, which I think is helpful information. I could just rename the diagnostic to `err_hlsl_binding_type_mismatch` from `err_hlsl_mismatching_register_type_and_resource_type`, but I think the existing message is better, what do you think?

My thoughts were that this error is for "the binding type is valid but it's applied to something that doesn't make sense". I found the proposed set of messages for this case to be kind of confusing and inconsistent, so the suggestion here is to not bother trying to classify what exactly the binding is being applied to, but just say that it doesn't make sense and explain what it *does* apply to.

> My issue with `err_hlsl_binding_type_invalid` replacing `err_hlsl_unsupported_register_type_and_variable_type`: It is important the diagnostic lists "for variable type x" instead of just declaring the register type invalid. This is because the diagnostic gets emitted not only when the register type is invalid (like "y"), but also when a completely invalid variable type is attempted to be bound (like RayQuery). We emit the diagnostic in this case, so it could be that "t" is used with RayQuery, which is a valid register type. It's just that the variable type is invalid. We could take your suggestion, but that would require adding a new diagnostic for cases where the variable type is all-around invalid. Thoughts?

If 't' is used with RayQuery we should just emit `err_hlsl_binding_type_mismatch` - RayQuery is not an SRV, end of story. Having a simple "This binding type is never valid" is simple and clear, so it makes sense to be its own thing.

> I can accept the rename to `err_hlsl_duplicate_register_annotation` I can accept the rename to `warn_hlsl_register_type_c_packoffset` I can accept both rewording suggestions.

Reply:
> > My issue with using `warn_hlsl_binding_type_mismatch` to replace `warn_hlsl_user_defined_type_missing_resource_type_member` and `warn_hlsl_user_defined_type_missing_basic_type`: 'c' is separate, because to my understanding, it isn't quite a register type, and it isn't binding a resource. So, the language used for 'c', "constant buffer data", should apply to register type 'b'. 'c' is for basic types that are to be placed in the $Globals constant buffer, which is a separate concept when compared to allocating a separate constant buffer, which is specified by 'b'. So I'm not quite sure 'c' is a "binding type". I could change the message to include a missing type for each of the 5 register types (t u b s c), but I think the separation of concepts is significant enough to warrant a separate diagnostic in the "c" case.
> 
> The way I see it, the 'b' binding type is for a constant buffer. The 'c' binding type is for "constant buffer data (in the $Globals constant buffer)". Given that 'b' bindings are _never_ valid on a VarDecl this separation seemed clean to me, and I think misuses of the 'b' type are the actual outlier here (which is covered by the `warn_hlsl_deprecated_register_type_b` diagnostic).

Ah, this isn't quite correct though, the 'b' binding is on a `VarDecl` here:

```hlsl
struct cbdata {
  float4 position;
};
ConstantBuffer<cbdata> cb1 : register(b);
```

even so, we have the special-cased separate diagnostic for when 'b' isn't correct so I don't think it changes much in practice.

Reply:
On further thought I decided to go ahead and take your suggestions, it makes diagnostic interpretation easier.